composer/utils/batch_helpers.py

```diff
@@ -0,0 +1,226 @@
+    ...     batch[1]['d'] = value
+    ...     return batch
+    ...
+    >>> batch_set([{'a':1, 'b':7},{'d':3}], value=20, set_fn=setter)
+    [{'a': 1, 'b': 7}, {'d': 20}]
+
+    Args:
+        batch (Any): An object that contains the input and label of the items in the batch.
+            Can be any abritrary type that user creates, but we assume some sort of
+            sequence (list, tuple, tensor, array), mapping (dictionary),
+            or attribute store (object with data members, namedtuple).
+        key (Any): A key to index into the batch.
+        value (Any): The value that batch[key] or batch.key gets set to.
+        set_fn (Callable): A user-specified function to do the setting. set_fn is optional if key and 
+            value are supplied. set_fn must return the updated batch.
+
+    Returns:
+        batch (Any): updated batch with value set at key.
+
+    Raises:
+        ValueError if:
+            * key and set_fn are both unset
+            * key and set_fn are both set
+    """
+    if key is None and set_fn is None:
+        raise ValueError("key or set_fn must be specified and neither were!")
+    if key is not None and set_fn is not None:
+        raise ValueError("key and set_fn were both set. Only one can be set!")
+    if set_fn:
+        return set_fn(batch, value)
+    if isinstance(key, Sequence) and not isinstance(key, str):
+        return _batch_set_multiple(batch, key, value)
+    else:
+        return _batch_set(batch, key, value)
+
+
+def _batch_set(batch: Any, key: Any, value: Any) -> Any:
+    """Sets a key value pair in a non-tuple batch."""
+    if isinstance(batch, tuple):
+        return _batch_set_tuple(batch, key, value)
+
+    exceptions = []
+    try:
+        # Check if one can do a __getitem__ before doing a __setitem__ because dicts can
+        # do __setitem__ for elements not in the dict and we do not want that.
+        # Note for defaultdict and Counter objects, just calling batch[key] for
+        # with a new keyword will create a new key, value pair in the object.
+        batch[key]
+        batch[key] = value
+
+    # The only acceptable TypeErrors are for an object that doesn't have a __setitem__ or a __getitem__,
+    # which is TypeError("... object does not support item assignment") and TypeError('.. object is not subscriptable')
+    except TypeError as e:
+        if 'object does not support item assignment' in str(e) or 'object is not subscriptable' in str(e):
+            exceptions.append(e)
+            pass
+        else:  # Other type errors should be raised.
+            raise e
+    else:
+        return batch
+
+    try:
+        # Make sure batch has key before setting it.
+        getattr(batch, key)
+        setattr(batch, key, value)
+
+    # If both (setattr or getattr) and __setitem__ raise exceptions then raise both of them.
+    except (AttributeError, TypeError) as e:
+        exceptions.append(e)
+        raise Exception(exceptions)
```

Comment:
Is it possible to raise a more specific exception? Perhaps a `TypeError`, `NotImplementedError`, `ValueError`, or a `RuntimeError`? I think any of these would be appropriate; don't have a strong opinion on which one. A more specific exception would make it possible for the caller to catch it.

In addition, is it necessary to re-raise the underlying exceptions? A bit torn on whether that would help the caller fix the issue. Thoughts about something like?

```suggestion
        raise RuntimeError(f"Unable to set key {key} to value {value} on batch {batch}. Please specify a custom set_fn, if necessary.")
```

Reply:
done here: 
bd3dce1